File: backend/app/observability/trace.py
from __future__ import annotations

from dataclasses import dataclass
from uuid import uuid4


@dataclass(frozen=True)
class TraceContext:
    """保存跨控制面、Worker 和证据事件传播的 trace/span 标识。"""
    trace_id: str
    span_id: str
    parent_span_id: str | None = None

    @classmethod
    def new(cls) -> "TraceContext":
        """创建新的根 trace 和 span。"""
        return cls(trace_id=f"tr_{uuid4().hex}", span_id=f"sp_{uuid4().hex}")

    def child(self) -> "TraceContext":
        """创建继承当前 trace、并以当前 span 为父级的新 span。"""
        return TraceContext(
            trace_id=self.trace_id,
            span_id=f"sp_{uuid4().hex}",
            parent_span_id=self.span_id,
        )

File: backend/src/observability/trace.ts
import { randomUUID } from "node:crypto";

/** 保存跨控制面、Worker 和证据事件传播的 trace/span 标识。 */
export class TraceContext {
  /** 保存 trace、span 和父 span 关系，供跨模块事件保持同一链路。 */
  constructor(
    readonly traceId: string,
    readonly spanId: string,
    readonly parentSpanId: string | null = null,
  ) {}
  /** 创建新的根 trace 和 span。 */
  static new(): TraceContext {
    return new TraceContext(
      `tr_${randomUUID().replaceAll("-", "")}`,
      `sp_${randomUUID().replaceAll("-", "")}`,
    );
  }
  /** 创建继承当前 trace 且以当前 span 为父级的新 span。 */
  child(): TraceContext {
    return new TraceContext(
      this.traceId,
      `sp_${randomUUID().replaceAll("-", "")}`,
      this.spanId,
    );
  }
}
